### PR TITLE
Macroscopic pml

### PR DESCRIPTION
**Pull request by RevathiJambunathan**
In this PR, the PML implementation is extended to account for macroscopic properties.

The input file  used to test this is attached below
[inputs_laser_mod.3d.txt](https://github.com/RevathiJambunathan/WarpX/files/5656945/inputs_laser_mod.3d.txt)

The test-case includes eps_r=4 for the material, and a laser in injected at an angle.
The current evolveM_dev code reflects the wave at the domain-boundary.

With this PR, now the waves are absorbed at the boundary.

with_macro_pml

### CODE REVIEW

File: Source/BoundaryConditions/PML.cpp [hunk 1]
```diff
@@ -11,6 +11,7 @@
 #include "Utils/WarpXConst.H"
 #include "Utils/WarpXAlgorithmSelection.H"
 #include "WarpX.H"
+//#include "FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.H"
```
Comment (ajnonaka) [inline Source/BoundaryConditions/PML.cpp]:
```suggestion
```

File: Source/BoundaryConditions/PML.cpp [hunk 2]
```diff
@@ -515,6 +517,42 @@ PML::PML (const BoxArray& grid_ba, const DistributionMapping& /*grid_dm*/,
...
+        // Initialize mu, permeability
+        if (macroscopic_properties->m_mu_s == "constant") {
+            pml_mu_fp->setVal(macroscopic_properties->m_mu);
+        } else if (macroscopic_properties->m_mu_s == "parse_mu_function") {
+            macroscopic_properties->InitializeMacroMultiFabUsingParser(pml_mu_fp.get(),
+                getParser(macroscopic_properties->m_mu_parser), lev);
+        }
+
+    }
+
```
Comment (ajnonaka) [inline Source/BoundaryConditions/PML.cpp:554]: General comment: the parser will read in based on (i,j,k) in the PML region.  Would it make sense to instead read the value from the parser ON the boundary instead of the location of the cell?  For AMR in the refined regions this would not be good.  I'm ok with leaving it as is, but then the material properites in the inputs file have to be carefully specified if a square of material is against the domain - we would have to extend this square out to infinity outside the domain, if that makes sense.

File: Source/FieldSolver/FiniteDifferenceSolver/FiniteDifferenceSolver.H [hunk 3]
```diff
@@ -128,12 +128,30 @@ class FiniteDifferenceSolver
...
+
+
 #ifdef WARPX_MAG_LLG
         void EvolveHPML ( std::array< amrex::MultiFab*, 3 > Hfield,
                       std::array< amrex::MultiFab*, 3 > const Efield,
                       amrex::Real const dt );
 #endif
 
+
+
```
Comment (ajnonaka) [inline Source/FieldSolver/FiniteDifferenceSolver/FiniteDifferenceSolver.H]:
```suggestion
```

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveEPML.cpp [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+    amrex::MultiFab* const eps_mf,
+    amrex::MultiFab* const mu_mf,
+    amrex::MultiFab* const sigma_mf ) {
+
+    Real c2 = PhysConst::c * PhysConst::c;
+
+#ifdef WARPX_MAG_LLG
+    c2 *= PhysConst::mu0;
+#endif
+
```
Comment (ajnonaka) [inline Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveEPML.cpp]:
```suggestion
```

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveEPML.cpp [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+            [=] AMREX_GPU_DEVICE (int i, int j, int k){
+                // Interpolate conductivity, sigma, to Ex position on the grid
+                amrex::Real const sigma_interp = CoarsenIO::Interp( sigma_arr, sigma_stag,
+                                           Ex_stag, macro_cr, i, j, k, scomp);
+                // Interpolated permittivity, epsilon, to Ex position on the grid
+                amrex::Real const epsilon_interp = CoarsenIO::Interp( eps_arr, epsilon_stag,
+                                           Ex_stag, macro_cr, i, j, k, scomp);
+                amrex::Real alpha = T_MacroAlgo::alpha( sigma_interp, epsilon_interp, dt);
+                amrex::Real beta = T_MacroAlgo::beta( sigma_interp, epsilon_interp, dt);
+
```
Comment (ajnonaka) [inline Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveEPML.cpp:193]: Comment: the non-macroscopic version doesn't have this alpha/beta part; was sigma assumed to be zero?

File: Source/BoundaryConditions/PML.cpp (hunk 2 above)
Comment (RevathiJambunathan) [reply Source/BoundaryConditions/PML.cpp:554]: YEs -- this is to be discussed.
The way it is right now -- it is easily extendable to pml inside the domain for MR.

But I agree that when we have a weird input with exponentially decreasing macroproperties, then for pml outside the domain, we need to take the boundary value.

For all our test-cases in the near future, this seems to be pretty linear or mostly even a jump condition with (zmid< and zmid>). Definitely addressing these corner cases is essential.
I am wondering if this can be done in a separate PR focusing only on corner cases for pml outside domain

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveEPML.cpp (hunk 4 above)
Comment (RevathiJambunathan) [reply Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveEPML.cpp]: oh thank you for catching this!

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveEPML.cpp (hunk 5 above)
Comment (RevathiJambunathan) [reply Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveEPML.cpp:193]: The alpha part comes from non-zero sigma
The beta comes from not using c^2, but using the factor function(sigma, epsilon).
This is similar to what we have in MacroscopicEvolveE.cpp

Yes, in the non-macrscopic version -- vacuum properties were assumed

File: Source/BoundaryConditions/PML.cpp (hunk 2 above)
Comment (ajnonaka) [reply Source/BoundaryConditions/PML.cpp:554]: I think it's ok to leave as it by now.  It matches up with all our short and medium-term use cases.
For advanced cases down the line with more complex functional forms this can be revisited.

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveEPML.cpp (hunk 5 above)
Comment (ajnonaka) [reply Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveEPML.cpp:193]: Thank you - I understand now.  Forgot that original EvolveE assumed sigma=0.

**Review by jackieyao0114 — APPROVED** [review]
Thank you for setting up the mu multifab in pml region! This is so helpful for updating MacroscopicEvolveHPML.

**Comment by RevathiJambunathan** [comment]
Thanks for the reviews @ajnonaka and @jackieyao0114
